pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,234 @@
+		return err
+	}
+	defer stream.CloseSend()
+	rqst, err := stream.Recv()
+	if err != nil {
+		glog.Errorf("Failed to receive stream at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	resp := new(watcherapi.RegistrationStatus)
+	if handlerCbkFn, ok := w.handlers[rqst.Type]; ok {
+		var versions []string
+		for _, version := range rqst.Version {
+			versions = append(versions, version)
+		}
+		// calls handler callback to verify registration request
+		if verifyErr := handlerCbkFn(rqst.Name, versions, socketPath); verifyErr != nil {
+			resp.PluginRegistered = false
+			resp.Error = fmt.Sprintf("Plugin registration failed with err: %v", verifyErr)
+		} else {
+			resp.PluginRegistered = true
+		}
+	} else {
+		resp.PluginRegistered = false
+		resp.Error = fmt.Sprintf("No handler found registered for plugin type: %s, socket: %s", rqst.Type, socketPath)
+	}
+	glog.V(2).Infof("send register resp: %v", resp)
+	if err := stream.Send(resp); err != nil {
+		glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+		return err
+	}
+
+	for {
+		_, err := stream.Recv()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			glog.Errorf("Recv failed. err: %v", err)
+			return err
+		}
+	}
+
+	if !resp.PluginRegistered {
+		glog.Errorf("Registration failed for plugin type: %s, name: %s, socket: %s, err: %s", rqst.Type, rqst.Name, socketPath, resp.Error)
+		return fmt.Errorf("%s", resp.Error)
+	}
+	glog.Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", rqst.Type, rqst.Name, socketPath)
+	return nil
+}
+
+// Start watches for the creation of plugin sockets at the path
+func (w *Watcher) Start() error {
+	glog.V(2).Infof("Plugin Watcher Start at %s", w.path)
+
+	// Creating the directory to be watched if it doesn't exist yet,
+	// and walks through the directory to discover the existing plugins.
+	if err := w.init(); err != nil {
+		return err
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+
+	err = watcher.Add(w.path)
+	if err != nil {
+		watcher.Close()
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+	w.watcher = watcher
+
+	go func(watcher *fsnotify.Watcher) {
+		for {
+			select {
+			case event := <-watcher.Events:
+				if event.Op&fsnotify.Create == fsnotify.Create {
+					w.registerPlugin(event.Name)
+				}
+				continue
+			case err := <-watcher.Errors:
+				//TODO: Handle errors by taking corrective measures
+				glog.Errorf("Watcher received error: %v", err)
+				continue
+
+			case <-w.stopCh:
+				watcher.Close()
+				w.stopCh <- "closed"
+				break
+			}
+		}
+	}(watcher)
+	return nil
+}
+
+// Stop stops probing the creation of plugin sockets at the path
+func (w *Watcher) Stop() error {
+	w.stopCh <- "done"
+	select {
+	case <-w.stopCh:
+	case <-time.After(5 * time.Second):
```

Comment:
5 seconds seem like a weird timeout considering the plugin dial timeout is 10s

Reply:
Dial timeout is on unix socket which is unrelated to this timeout which is an intra process channel. What is weird in 5 sec?

Reply:
goroutine#1) fsnotify recieves socket creation event
goroutine#1) The plugin_watcher starts dialing the socket (but it will timeout after 10s)
goroutine#2) Stop event is recieved
goroutine#2) Timeout after 5 seconds

If the dial timeout was better tuned the Stop timeout would not have been triggered here.
Additionally as mentioned higher you should be using a context that can be cancelled in the Stop function.